src/client/types.ts

```diff
@@ -232,20 +232,48 @@ export interface MetadataApiRetrieveStatus {
 // Client options
 // ------------------------------------------------
 
+export interface PackageOptions {
+  /**
+   * The name of the package to retrieve.
+   */
+  name: string;
+  /**
+   * The directory where the retrieved package source should be
+   * converted. If this is not specified the directory will
+   * default to `<process.cwd()>/PackageOptions.name`.
+   */
+  outputDir?: SourcePath;
+}
+
+export interface RetrieveExtractOptions {
+  /**
+   * Top most directory within the retrieved zip file.
+   * E.g., `unpackaged` for unpackaged source, or the name of the
+   * package for retrieved package source.
+   */
+  zipTreeLocation: string;
+  /**
+   * The directory where the retrieved source should be converted.
+   * This is `RetrieveOptions.output` for unpackaged source, and
+   * `PackageOptions.outputDir` for packaged source.
+   */
+  outputDir: SourcePath;
+}
+
 export interface RetrieveOptions {
   /**
    * The directory to retrieve components to. If `merge: true`, components are only
    * retrieved to `output` if there wasn't a component to merge with.
    */
   output: SourcePath;
   /**
-   * Whether or not to merge and replace input components with the retrieved versions
+   * Whether or not to merge and replace input components with the retrieved versions.
    */
   merge?: boolean;
   /**
-   * A list of package names to retrieve
+   * A list of package names to retrieve, or package names and their retrieval locations.
    */
-  packageNames?: string[];
+  packages?: string[] | PackageOptions[];
```

Comment:
For the sake of grokkability maybe `packageOptions` is a better name for this. Also, semantically it looks cleaner to define this as a union type `type PackageOptions = string | { ... }`.

Reply:
A PackageOption is `{ name: string; outputDir?: string }` while `string[]` is an array of package names, which doesn't seem like an option to me.  The internal code needs to distinguish between the 2 so they really aren't 1 thing.  I.e., "if packages is string[] then cast to string[] and do this, but if not then cast to PackageOption[] and do this."  If it's all a `PackageOption[]` then we'd need to come up with another subtype to refer to `{ name: string; outputDir: string }[]`.

The js doc makes it pretty clear what the types mean but we could do this and it would cause less refactoring:
type PackageNames = string[]
type PackageOptions = { name: string; outputDir: string }[]
...
packages?: PackageNames | PackageOptions;